lms/services/assignment.py

```diff
@@ -10,55 +12,210 @@ def __init__(self, db):
+                tool_consumer_instance_guid, ext_lti_assignment_id
+            )
 
-        return assignment.document_url if assignment else None
+        # We have both ext_lti_assignment_id and resource_link_id, canvas launch.
+        return self._get_for_canvas_launch(
+            tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+        )
+
+    def exists(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ) -> bool:
+        try:
+            self.get(
+                tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+            )
+        except (NoResultFound, ValueError):
+            return False
+        else:
+            return True
+
+    def create(
+        self,
+        tool_consumer_instance_guid,
+        document_url,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ):
+        assignment = Assignment(
+            tool_consumer_instance_guid=tool_consumer_instance_guid,
+            document_url=document_url,
+            resource_link_id=resource_link_id,
+            ext_lti_assignment_id=ext_lti_assignment_id,
+        )
+        self._db.add(assignment)
+        return assignment
 
     def set_document_url(
-        self, tool_consumer_instance_guid, resource_link_id, document_url
+        self,
+        tool_consumer_instance_guid,
+        document_url,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
     ):
         """
-        Save the given document_url.
+        Save the given document_url in an existing assignment or create new one.
 
         Set the document_url for the assignment that matches
-        tool_consumer_instance_guid and resource_link_id. Any existing document
-        URL for this assignment will be overwritten.
-        """
-        assignment = self.get(tool_consumer_instance_guid, resource_link_id)
+        tool_consumer_instance_guid and resource_link_id/ext_lti_assignment_id
+        or create a new one if none exist on the DB.
 
-        if assignment:
+        Any existing document URL for this assignment will be overwritten.
+        """
+        try:
+            assignment = self.get(
+                tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+            )
             assignment.document_url = document_url
-        else:
-            self._db.add(
-                Assignment(
-                    document_url=document_url,
-                    resource_link_id=resource_link_id,
-                    tool_consumer_instance_guid=tool_consumer_instance_guid,
-                )
+        except NoResultFound:
+            assignment = self.create(
+                document_url=document_url,
+                resource_link_id=resource_link_id,
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                ext_lti_assignment_id=ext_lti_assignment_id,
```

Comment:
I think we can refactor this a little so that there's only one line in the `try` and so that it's clear that the intention is to set the `document_url` on either a pre-existing or a newly-created assignment:

```python
try:
    assignment = self.get(...)  # Get the existing assignment if there is one.
except NoResultFound:
    # If there's no existing assignment create a new one.
    assignment = Assignment(...)  # Here pass everything *except* document_url to Assignment().
    self._db.add(assignment)

assignment.document_url = document_url
```

This `except` is taking advantage of the fact that SQLAlchemy's default interface lets you call `Assignment(...)` with any kwargs, or even no kwargs, so you can call it without `document_url`. I think you actually probably don't want the `AssignmentService.create()` method: it's taking two lines of code in the `except` and replacing them with one line in the `except` plus a 14-line `create()` method and it's less flexible because when calling `create()` you can't omit the `tool_consumer_instance_guid` or `document_url` and then set them later like you can when calling `Assignment()` directly.